frontend/lib/src/components/elements/ArrowVegaLiteChart/ArrowVegaLiteChart.tsx

```diff
@@ -269,21 +367,164 @@ export class ArrowVegaLiteChart extends PureComponent<
+      // the element ID must be set.
+      return
+    }
+
+    // Try to load the previous state of the chart from the element state.
+    // This is useful to restore the selection state when the component is re-mounted
+    // or when its put into fullscreen mode.
+    const viewState = widgetMgr.getElementState(
+      this.props.element.id,
+      "viewState"
+    )
+    if (notNullOrUndefined(viewState)) {
+      try {
+        this.vegaView = this.vegaView.setState(viewState)
+      } catch (e) {
+        logWarning("Failed to restore view state", e)
+      }
+    }
+
+    // Add listeners for all selection events. Find out more here:
+    // https://vega.github.io/vega/docs/api/view/#view_addSignalListener
+    element.selectionMode.forEach((param, _index) => {
+      this.vegaView?.addSignalListener(
+        param,
+        debounce(DEBOUNCE_TIME_MS, (name: string, value: SignalValue) => {
+          // Store the current chart selection state with the widget manager so that it
+          // can be used for restoring the state when the component unmounted and
+          // created again. This can happen when elements are added before it within
+          // the delta path.
+          const viewState = this.vegaView?.getState({
+            // There are also `signals` data, but I believe its
+            // not relevant for restoring the selection state.
+            data: (_name?: string, _operator?: any) => {
+              // Vega lite stores the selection state in a <param name>_store parameter
+              // under `data` that can be retrieved via the getState method.
+              // https://vega.github.io/vega/docs/api/view/#view_getState
+              const stateNames = element.selectionMode.map(
+                (param, _idx) => `${param}_store`
+              )
+              return _name ? stateNames.includes(_name) : false
+            },
+            // Don't include subcontext data since it will lead to exceptions
+            // when loading the state.
+            recurse: false,
+          })
+
+          if (notNullOrUndefined(viewState)) {
+            widgetMgr.setElementState(element.id, "viewState", viewState)
+          }
+
+          // If selection encodings are correctly specified, vega-lite will return
+          // a list of selected points within the vlPoint.or property:
+          // https://github.com/vega/altair/blob/f1b4e2c84da2fba220022c8a285cc8280f824ed8/altair/utils/selection.py#L50
+          // We want to just return this list of points instead of the entire object
+          // since the other parts of the selection object are not useful.
+          let processedSelection = value
+          if ("vlPoint" in value && "or" in value.vlPoint) {
+            processedSelection = value.vlPoint.or
+          }
+
+          // Get the current widget state
+          const currentWidgetState = JSON.parse(
+            widgetMgr.getStringValue(element as WidgetInfo) || "{}"
+          )
```

Comment:
we seem to set the widget state [above](https://github.com/streamlit/streamlit/pull/8302/files#diff-92f0e57b8f2d70900bd5b548ed97c7fb579f2880d48faaea8b7cfac004d479c0R443); so shouldn't we be able to just use `viewState` here?

Reply:
These are two slightly different things. Above we are setting an element state that is purely used in the frontend to reconstruct state after unmounting. The state here is whats actually synced with the backend, but this state is not used for reconstructing the view state.

Reply:
I added some more comments